diff --git a/common.mk b/common.mk
index e9b24bc..cd130ef 100644
--- a/common.mk
+++ b/common.mk
@@ -119,11 +119,6 @@ PRODUCT_PACKAGES += \
 PRODUCT_COPY_FILES += \
     frameworks/native/data/etc/android.hardware.touchscreen.multitouch.jazzhand.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.touchscreen.multitouch.jazzhand.xml
 
-# Doze
-PRODUCT_PACKAGES += \
-    OplusDoze \
-    OplusDozeResCommon
-
 # DRM
 PRODUCT_PACKAGES += \
     android.hardware.drm-service.clearkey \
@@ -250,7 +245,7 @@ PRODUCT_PACKAGES += \
 
 # Sensors
 PRODUCT_PACKAGES += \
-    android.hardware.sensors-service.multihal \
+    android.hardware.sensors-service.oplus-multihal \
     sensors.oplus
 
 PRODUCT_COPY_FILES += \
diff --git a/overlay/OplusDozeResCommon/res/values/config.xml b/overlay-lineage/hardware/oplus/doze/res/values/config.xml
similarity index 86%
rename from overlay/OplusDozeResCommon/res/values/config.xml
rename to overlay-lineage/hardware/oplus/doze/res/values/config.xml
index 50c067b..b9c1fef 100644
--- a/overlay/OplusDozeResCommon/res/values/config.xml
+++ b/overlay-lineage/hardware/oplus/doze/res/values/config.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!--
-     SPDX-FileCopyrightText: The LineageOS Project
+     Copyright (C) 2021-2022 The LineageOS Project
      SPDX-License-Identifier: Apache-2.0
 -->
 <resources>
diff --git a/overlay/OplusDozeResCommon/Android.bp b/overlay/OplusDozeResCommon/Android.bp
deleted file mode 100644
index 3c57739..0000000
--- a/overlay/OplusDozeResCommon/Android.bp
+++ /dev/null
@@ -1,9 +0,0 @@
-//
-// SPDX-FileCopyrightText: The LineageOS Project
-// SPDX-License-Identifier: Apache-2.0
-//
-
-runtime_resource_overlay {
-    name: "OplusDozeResCommon",
-    system_ext_specific: true,
-}
diff --git a/overlay/OplusDozeResCommon/AndroidManifest.xml b/overlay/OplusDozeResCommon/AndroidManifest.xml
deleted file mode 100644
index 84aae14..0000000
--- a/overlay/OplusDozeResCommon/AndroidManifest.xml
+++ /dev/null
@@ -1,11 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!--
-     SPDX-FileCopyrightText: The LineageOS Project
-     SPDX-License-Identifier: Apache-2.0
--->
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-          package="org.lineageos.settings.doze.overlay">
-    <overlay android:targetPackage="org.lineageos.settings.doze"
-             android:isStatic="true"
-             android:priority="300" />
-</manifest>
diff --git a/vendor.prop b/vendor.prop
index 51d6f7c..563b9dd 100644
--- a/vendor.prop
+++ b/vendor.prop
@@ -163,6 +163,7 @@ ro.telephony.default_network=33,33
 
 # Sensors
 persist.vendor.sensors.allow_non_default_discovery=true
+persist.vendor.sensors.debug.hal=0
 persist.vendor.sensors.on_change_sample_period=true
 persist.vendor.sensors.sync_request=true
 
