diff --git a/core/java/com/oplus/theme/OplusThemeUtil.java b/core/java/com/oplus/theme/OplusThemeUtil.java
new file mode 100644
index 000000000000..cea1cfd31b83
--- /dev/null
+++ b/core/java/com/oplus/theme/OplusThemeUtil.java
@@ -0,0 +1,9 @@
+package com.oplus.theme;
+
+import android.compat.annotation.UnsupportedAppUsage;
+
+public class OplusThemeUtil {
+    /** @hide */
+    @UnsupportedAppUsage
+    public static String CUSTOM_THEME_PATH = "/data/theme/com.oplus.camera";
+}
diff --git a/core/java/com/oplus/util/OplusTypeCastingHelper.java b/core/java/com/oplus/util/OplusTypeCastingHelper.java
new file mode 100644
index 000000000000..4819e9f20b72
--- /dev/null
+++ b/core/java/com/oplus/util/OplusTypeCastingHelper.java
@@ -0,0 +1,10 @@
+package com.oplus.util;
+
+public final class OplusTypeCastingHelper {
+    public static <T> T typeCasting(Class<T> type, Object object) {
+        if (object != null && type.isInstance(object)) {
+            return type.cast(object);
+        }
+        return null;
+    }
+}
diff --git a/core/java/android/hardware/camera2/impl/CameraMetadataNative.java b/core/java/android/hardware/camera2/impl/CameraMetadataNative.java
index 5da8d5436274..eb1268657a6f 100644
--- a/core/java/android/hardware/camera2/impl/CameraMetadataNative.java
+++ b/core/java/android/hardware/camera2/impl/CameraMetadataNative.java
@@ -2352,6 +2352,18 @@ public class CameraMetadataNative implements Parcelable {
     private static native int nativeGetTypeFromTag(int tag, long vendorId)
             throws IllegalArgumentException;
 
+    private synchronized byte[] nativeReadValues(int tag) {
+        return nativeReadValues(tag, mMetadataPtr);
+    }
+
+    private synchronized int nativeGetTypeFromTagLocal(int tag) {
+        return nativeGetTypeFromTagLocal(mMetadataPtr, tag);
+    }
+
+    private synchronized int nativeGetTagFromKeyLocal(String keyname) {
+        return nativeGetTagFromKeyLocal(mMetadataPtr, keyname);
+    }
+
     /**
      * <p>Perform a 0-copy swap of the internal metadata with another object.</p>
      *
