diff --git a/arch/arm64/boot/dts/vendor/qcom/blair.dtsi b/arch/arm64/boot/dts/vendor/qcom/blair.dtsi
index 56c2de2958f5..07d9f426a828 100644
--- a/arch/arm64/boot/dts/vendor/qcom/blair.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/blair.dtsi
@@ -34,14 +34,6 @@
 
 	reserved_memory: reserved-memory { };
 
-	mem-offline {
-		compatible = "qcom,mem-offline";
-		offline-sizes = <0x1 0x40000000 0x0 0x40000000>,
-				<0x1 0xc0000000 0x0 0x80000000>,
-				<0x2 0xc0000000 0x1 0x40000000>;
-		granule = <512>;
-	};
-
 	aliases {
 		ufshc1 = &ufshc_mem; /* Embedded  UFS Slot */
 
diff --git a/arch/arm64/boot/dts/vendor/qcom/holi.dtsi b/arch/arm64/boot/dts/vendor/qcom/holi.dtsi
index 9b717d6d1f6c..b24d0271d595 100644
--- a/arch/arm64/boot/dts/vendor/qcom/holi.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/holi.dtsi
@@ -30,14 +30,6 @@
 	memory { device_type = "memory"; reg = <0 0 0 0>; };
 	reserved_memory: reserved-memory { };
 
-	mem-offline {
-		compatible = "qcom,mem-offline";
-		offline-sizes = <0x1 0x40000000 0x0 0x40000000>,
-				<0x1 0xc0000000 0x0 0x80000000>,
-				<0x2 0xc0000000 0x1 0x40000000>;
-		granule = <512>;
-	};
-
 	aliases {
 		ufshc1 = &ufshc_mem; /* Embedded  UFS Slot */
 
diff --git a/arch/arm64/boot/dts/vendor/qcom/lahaina.dtsi b/arch/arm64/boot/dts/vendor/qcom/lahaina.dtsi
index 5327f3a02516..6b6d2a733bbb 100644
--- a/arch/arm64/boot/dts/vendor/qcom/lahaina.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/lahaina.dtsi
@@ -40,15 +40,6 @@
 	memory { device_type = "memory"; reg = <0 0 0 0>; };
 	reserved_memory: reserved-memory { };
 
-	mem-offline {
-		compatible = "qcom,mem-offline";
-		offline-sizes = <0x1 0x40000000 0x0 0x40000000>,
-				<0x1 0xc0000000 0x0 0x80000000>,
-				<0x2 0xc0000000 0x1 0x40000000>;
-		granule = <512>;
-		mboxes = <&qmp_aop 0>;
-	};
-
 	aliases {
 		ufshc1 = &ufshc_mem; /* Embedded UFS Slot */
 		sdhc1 = &sdhc_2; /* SDC2 SD card slot */
diff --git a/arch/arm64/boot/dts/vendor/qcom/qrb2210-rb1.dtsi b/arch/arm64/boot/dts/vendor/qcom/qrb2210-rb1.dtsi
index 6e6e4410b3d8..bb43f251727e 100644
--- a/arch/arm64/boot/dts/vendor/qcom/qrb2210-rb1.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/qrb2210-rb1.dtsi
@@ -22,14 +22,6 @@
 
 	memory { device_type = "memory"; reg = <0 0 0 0>; };
 
-	mem-offline {
-		compatible = "qcom,mem-offline";
-		offline-sizes = <0x1 0x40000000 0x0 0x40000000>,
-				<0x1 0xc0000000 0x0 0x80000000>,
-				<0x2 0xc0000000 0x1 0x40000000>;
-		granule = <512>;
-	};
-
 	aliases {
 		serial0 = &qupv3_se4_2uart;
 		hsuart0 = &qupv3_se3_4uart;
diff --git a/arch/arm64/boot/dts/vendor/qcom/scuba-iot.dtsi b/arch/arm64/boot/dts/vendor/qcom/scuba-iot.dtsi
index 1745280ef456..072024e997e2 100644
--- a/arch/arm64/boot/dts/vendor/qcom/scuba-iot.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/scuba-iot.dtsi
@@ -15,14 +15,6 @@
 
 	memory { device_type = "memory"; reg = <0 0 0 0>; };
 
-	mem-offline {
-		compatible = "qcom,mem-offline";
-		offline-sizes = <0x1 0x40000000 0x0 0x40000000>,
-				<0x1 0xc0000000 0x0 0x80000000>,
-				<0x2 0xc0000000 0x1 0x40000000>;
-		granule = <512>;
-	};
-
 	aliases {
 		swr0 = &swr0;
 		swr1 = &swr1;
diff --git a/arch/arm64/boot/dts/vendor/qcom/shima.dtsi b/arch/arm64/boot/dts/vendor/qcom/shima.dtsi
index 1bd9b8c51413..e771c59525dc 100644
--- a/arch/arm64/boot/dts/vendor/qcom/shima.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/shima.dtsi
@@ -39,15 +39,6 @@
 	memory { device_type = "memory"; reg = <0 0 0 0>; };
 	reserved_memory: reserved-memory { };
 
-	mem-offline {
-		compatible = "qcom,mem-offline";
-		offline-sizes = <0x1 0x40000000 0x0 0x40000000>,
-			<0x1 0xc0000000 0x0 0x80000000>,
-			<0x2 0xc0000000 0x1 0x40000000>;
-		granule = <512>;
-		mboxes = <&qmp_aop 0>;
-	};
-
 	aliases {
 			ufshc1 = &ufshc_mem; /* Embedded UFS Slot */
 			sdhc0 = &sdhc_1; /*SDC1 eMMC slot*/
diff --git a/arch/arm64/boot/dts/vendor/qcom/yupik.dtsi b/arch/arm64/boot/dts/vendor/qcom/yupik.dtsi
index e9e5c07a5a38..982c87e283c2 100644
--- a/arch/arm64/boot/dts/vendor/qcom/yupik.dtsi
+++ b/arch/arm64/boot/dts/vendor/qcom/yupik.dtsi
@@ -42,15 +42,6 @@
 
 	reserved_memory: reserved-memory { };
 
-	mem-offline {
-		compatible = "qcom,mem-offline";
-		offline-sizes = <0x1 0x40000000 0x0 0x40000000>,
-			<0x1 0xc0000000 0x0 0x80000000>,
-			<0x2 0xc0000000 0x1 0x40000000>;
-		granule = <512>;
-		mboxes = <&qmp_aop 0>;
-	};
-
 	aliases {
 		ufshc1 = &ufshc_mem; /* Embedded UFS Slot */
 		sdhc0 = &sdhc_1; /*SDC1 eMMC slot*/
diff --git a/arch/arm64/configs/vendor/holi_QGKI.config b/arch/arm64/configs/vendor/holi_QGKI.config
index c33b461c436c..a962fc6e0840 100644
--- a/arch/arm64/configs/vendor/holi_QGKI.config
+++ b/arch/arm64/configs/vendor/holi_QGKI.config
@@ -35,13 +35,7 @@ CONFIG_MSM_GLOBAL_SYNX=y
 CONFIG_MSM_CVP=y
 CONFIG_IIO=y
 CONFIG_ION_POOL_AUTO_REFILL=y
-CONFIG_HOTPLUG_SIZE_BITS=29
-CONFIG_MEMORY_HOTPLUG_DEFAULT_ONLINE=y
-CONFIG_MEMORY_HOTPLUG_MOVABLE_NODE=y
-CONFIG_MEMORY_HOTREMOVE=y
 CONFIG_QCOM_MINIDUMP=y
-CONFIG_QCOM_MEM_OFFLINE=y
-CONFIG_OVERRIDE_MEMORY_LIMIT=y
 CONFIG_VM_EVENT_COUNT_CLEAN_PAGE_RECLAIM=y
 CONFIG_BALANCE_ANON_FILE_RECLAIM=y
 CONFIG_PRIORITIZE_OOM_TASKS=y
diff --git a/arch/arm64/configs/vendor/lahaina-qgki_defconfig b/arch/arm64/configs/vendor/lahaina-qgki_defconfig
index 44b37a247f0a..2027c8e48f74 100644
--- a/arch/arm64/configs/vendor/lahaina-qgki_defconfig
+++ b/arch/arm64/configs/vendor/lahaina-qgki_defconfig
@@ -755,9 +755,6 @@ CONFIG_QCOM_MINIDUMP=y
 CONFIG_QCOM_MICRODUMP=y
 CONFIG_QCOM_SECURE_BUFFER=y
 CONFIG_QCOM_GLINK_PKT=y
-CONFIG_MSM_PERFORMANCE=y
-CONFIG_QTI_PLH=y
-CONFIG_MSM_PERFORMANCE_QGKI=y
 CONFIG_QSEE_IPC_IRQ_BRIDGE=y
 CONFIG_QCOM_SMP2P_SLEEPSTATE=y
 CONFIG_SENSORS_SSC=y
@@ -770,7 +767,6 @@ CONFIG_QTI_ALTMODE_GLINK=y
 CONFIG_QTI_PMIC_PON_LOG=y
 CONFIG_MSM_CDSP_LOADER=y
 CONFIG_QGKI=y
-CONFIG_QTI_PLH_SCMI_CLIENT=y
 CONFIG_QTI_HW_MEMLAT=y
 CONFIG_QTI_HW_MEMLAT_SCMI_CLIENT=y
 CONFIG_QTI_HW_MEMLAT_LOG=y
diff --git a/arch/arm64/configs/vendor/lahaina_QGKI.config b/arch/arm64/configs/vendor/lahaina_QGKI.config
index e3e057f67630..e7715d5c2ba0 100755
--- a/arch/arm64/configs/vendor/lahaina_QGKI.config
+++ b/arch/arm64/configs/vendor/lahaina_QGKI.config
@@ -45,13 +45,7 @@ CONFIG_ARM_QCOM_LPM_CPUIDLE=y
 CONFIG_MSM_GLOBAL_SYNX=y
 CONFIG_MSM_CVP=y
 CONFIG_ION_POOL_AUTO_REFILL=y
-CONFIG_HOTPLUG_SIZE_BITS=29
-CONFIG_MEMORY_HOTPLUG_DEFAULT_ONLINE=y
-CONFIG_MEMORY_HOTPLUG_MOVABLE_NODE=y
-CONFIG_MEMORY_HOTREMOVE=y
 CONFIG_QCOM_MINIDUMP=y
-CONFIG_QCOM_MEM_OFFLINE=y
-CONFIG_OVERRIDE_MEMORY_LIMIT=y
 CONFIG_QCOM_MEM_BUF=y
 CONFIG_VM_EVENT_COUNT_CLEAN_PAGE_RECLAIM=y
 CONFIG_BALANCE_ANON_FILE_RECLAIM=y
diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 40bf624d0cb1..e31b72a674e0 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -507,7 +507,10 @@ static int selinux_is_genfs_special_handling(struct super_block *sb)
 		!strcmp(sb->s_type->name, "rootfs") ||
 		(selinux_policycap_cgroupseclabel() &&
 		 (!strcmp(sb->s_type->name, "cgroup") ||
-		  !strcmp(sb->s_type->name, "cgroup2")));
+		  !strcmp(sb->s_type->name, "cgroup2"))) ||
+		// Android: remove functionfs policycap check due to
+		// ABI breakage with policycap array.
+		!strcmp(sb->s_type->name, "functionfs");
 }
 
 static int selinux_is_sblabel_mnt(struct super_block *sb)
@@ -759,7 +762,11 @@ static int selinux_set_mnt_opts(struct super_block *sb,
 	    !strcmp(sb->s_type->name, "tracefs") ||
 	    !strcmp(sb->s_type->name, "binder") ||
 	    !strcmp(sb->s_type->name, "bpf") ||
-	    !strcmp(sb->s_type->name, "pstore"))
+	    !strcmp(sb->s_type->name, "pstore") ||
+	    !strcmp(sb->s_type->name, "securityfs") ||
+	    // Android: remove functionfs policycap check due to
+	    // ABI breakage with policycap array.
+	    !strcmp(sb->s_type->name, "functionfs"))
 		sbsec->flags |= SE_SBGENFS;
 
 	if (!strcmp(sb->s_type->name, "sysfs") ||
diff --git a/techpack/display/msm/sde/sde_crtc.c b/techpack/display/msm/sde/sde_crtc.c
index cc0425590603..5cafbf8fc6ef 100644
--- a/techpack/display/msm/sde/sde_crtc.c
+++ b/techpack/display/msm/sde/sde_crtc.c
@@ -239,6 +239,47 @@ static int _sde_debugfs_fps_status(struct inode *inode, struct file *file)
 }
 #endif
 
+static ssize_t early_wakeup_store(struct device *device,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	struct drm_crtc *crtc;
+	struct sde_crtc *sde_crtc;
+	struct msm_drm_private *priv;
+	u32 crtc_id;
+	bool trigger;
+
+	if (!device || !buf || !count) {
+		SDE_ERROR("invalid input param(s)\n");
+		return -EINVAL;
+	}
+
+	if (kstrtobool(buf, &trigger) < 0)
+		return -EINVAL;
+
+	if (!trigger)
+		return count;
+
+	crtc = dev_get_drvdata(device);
+	if (!crtc || !crtc->dev || !crtc->dev->dev_private) {
+		SDE_ERROR("invalid crtc\n");
+		return -EINVAL;
+	}
+
+	sde_crtc = to_sde_crtc(crtc);
+	priv = crtc->dev->dev_private;
+
+	crtc_id = drm_crtc_index(crtc);
+	if (crtc_id >= ARRAY_SIZE(priv->disp_thread)) {
+		SDE_ERROR("invalid crtc index[%d]\n", crtc_id);
+		return -EINVAL;
+	}
+
+	kthread_queue_work(&priv->disp_thread[crtc_id].worker,
+			&sde_crtc->early_wakeup_work);
+
+	return count;
+}
+
 static ssize_t fps_periodicity_ms_store(struct device *device,
 		struct device_attribute *attr, const char *buf, size_t count)
 {
@@ -425,12 +466,14 @@ static DEVICE_ATTR_RO(vsync_event);
 static DEVICE_ATTR_RO(measured_fps);
 static DEVICE_ATTR_RW(fps_periodicity_ms);
 static DEVICE_ATTR_RO(retire_frame_event);
+static DEVICE_ATTR_WO(early_wakeup);
 
 static struct attribute *sde_crtc_dev_attrs[] = {
 	&dev_attr_vsync_event.attr,
 	&dev_attr_measured_fps.attr,
 	&dev_attr_fps_periodicity_ms.attr,
 	&dev_attr_retire_frame_event.attr,
+	&dev_attr_early_wakeup.attr,
 	NULL
 };
 
@@ -7060,6 +7103,40 @@ void sde_crtc_cancel_delayed_work(struct drm_crtc *crtc)
 	SDE_EVT32(DRMID(crtc), idle_status, cache_status);
 }
 
+/*
+ * __sde_crtc_early_wakeup_work - trigger early wakeup from user space
+ */
+static void __sde_crtc_early_wakeup_work(struct kthread_work *work)
+{
+	struct sde_crtc *sde_crtc = container_of(work, struct sde_crtc,
+				early_wakeup_work);
+	struct drm_crtc *crtc;
+	struct drm_device *dev;
+	struct msm_drm_private *priv;
+	struct sde_kms *sde_kms;
+
+	if (!sde_crtc) {
+		SDE_ERROR("invalid sde crtc\n");
+		return;
+	}
+
+	if (!sde_crtc->enabled) {
+		SDE_INFO("sde crtc is not enabled\n");
+		return;
+	}
+
+	crtc = &sde_crtc->base;
+	dev = crtc->dev;
+	if (!dev) {
+		SDE_ERROR("invalid drm device\n");
+		return;
+	}
+
+	priv = dev->dev_private;
+	sde_kms = to_sde_kms(priv->kms);
+	sde_kms_trigger_early_wakeup(sde_kms, crtc);
+}
+
 /* initialize crtc */
 struct drm_crtc *sde_crtc_init(struct drm_device *dev, struct drm_plane *plane)
 {
@@ -7156,6 +7233,8 @@ struct drm_crtc *sde_crtc_init(struct drm_device *dev, struct drm_plane *plane)
 					__sde_crtc_idle_notify_work);
 	kthread_init_delayed_work(&sde_crtc->static_cache_read_work,
 			__sde_crtc_static_cache_read_work);
+	kthread_init_work(&sde_crtc->early_wakeup_work,
+					__sde_crtc_early_wakeup_work);
 
 	SDE_DEBUG("crtc=%d new_llcc=%d, old_llcc=%d\n",
 		crtc->base.id,
diff --git a/techpack/display/msm/sde/sde_crtc.h b/techpack/display/msm/sde/sde_crtc.h
index c4f05e2192b9..af06b7848136 100644
--- a/techpack/display/msm/sde/sde_crtc.h
+++ b/techpack/display/msm/sde/sde_crtc.h
@@ -278,6 +278,7 @@ struct sde_crtc_misr_info {
  * @misr_frame_count  : misr frame count provided by client
  * @misr_data     : store misr data before turning off the clocks.
  * @idle_notify_work: delayed worker to notify idle timeout to user space
+ * @early_wakeup_work: work to trigger early wakeup
  * @power_event   : registered power event handle
  * @cur_perf      : current performance committed to clock/bandwidth driver
  * @plane_mask_old: keeps track of the planes used in the previous commit
@@ -361,6 +362,7 @@ struct sde_crtc {
 	bool misr_reconfigure;
 	u32 misr_frame_count;
 	struct kthread_delayed_work idle_notify_work;
+	struct kthread_work early_wakeup_work;
 
 	struct sde_power_event *power_event;
 
diff --git a/techpack/display/msm/sde/sde_encoder.c b/techpack/display/msm/sde/sde_encoder.c
index 5b191d99773e..4c9c2e6087ee 100644
--- a/techpack/display/msm/sde/sde_encoder.c
+++ b/techpack/display/msm/sde/sde_encoder.c
@@ -6220,3 +6220,27 @@ bool sde_encoder_is_disabled(struct drm_encoder *drm_enc)
 	return (phys->enable_state == SDE_ENC_DISABLED);
 }
 #endif
+
+void sde_encoder_trigger_early_wakeup(struct drm_encoder *drm_enc)
+{
+	struct sde_encoder_virt *sde_enc = NULL;
+	struct msm_drm_private *priv = NULL;
+
+	priv = drm_enc->dev->dev_private;
+	sde_enc = to_sde_encoder_virt(drm_enc);
+	if (!sde_enc->crtc || (sde_enc->crtc->index
+			>= ARRAY_SIZE(priv->disp_thread))) {
+		SDE_DEBUG_ENC(sde_enc,
+			"invalid cached CRTC: %d or crtc index: %d\n",
+			sde_enc->crtc == NULL,
+			sde_enc->crtc ? sde_enc->crtc->index : -EINVAL);
+		return;
+	}
+
+	if (sde_enc->rc_state == SDE_ENC_RC_STATE_IDLE) {
+		SDE_ATRACE_BEGIN("sde_encoder_resource_control");
+		sde_encoder_resource_control(drm_enc,
+					SDE_ENC_RC_EVENT_EARLY_WAKEUP);
+		SDE_ATRACE_END("sde_encoder_resource_control");
+	}
+}
diff --git a/techpack/display/msm/sde/sde_encoder.h b/techpack/display/msm/sde/sde_encoder.h
index 3c0eb14b931d..f3fe86d6314b 100644
--- a/techpack/display/msm/sde/sde_encoder.h
+++ b/techpack/display/msm/sde/sde_encoder.h
@@ -706,4 +706,11 @@ void sde_encoder_disable_autorefresh_handler(struct drm_encoder *drm_enc);
  */
 bool sde_encoder_is_disabled(struct drm_encoder *drm_enc);
 #endif
+
+/**
+ * sde_encoder_trigger_early_wakeup - trigger early wake up
+ * @drm_enc:    Pointer to drm encoder structure
+ */
+void sde_encoder_trigger_early_wakeup(struct drm_encoder *drm_enc);
+
 #endif /* __SDE_ENCODER_H__ */
diff --git a/techpack/display/msm/sde/sde_hw_color_proc_v4.c b/techpack/display/msm/sde/sde_hw_color_proc_v4.c
index ef11a0fe608c..919c9ffbda6a 100644
--- a/techpack/display/msm/sde/sde_hw_color_proc_v4.c
+++ b/techpack/display/msm/sde/sde_hw_color_proc_v4.c
@@ -1,7 +1,9 @@
 // SPDX-License-Identifier: GPL-2.0-only
 /*
  * Copyright (c) 2017-2021, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2018, Pal Zoltan Illes (tbalden) - kcal rgb
  */
+#include <linux/moduleparam.h>
 #include <drm/msm_drm_pp.h>
 #include "sde_hw_color_proc_common_v4.h"
 #include "sde_hw_color_proc_v4.h"
@@ -13,6 +15,22 @@ extern struct dsi_display *get_main_display(void);
 #endif
 #include "sde_trace.h"
 
+static unsigned short kcal_red = 256;
+static unsigned short kcal_green = 256;
+static unsigned short kcal_blue = 256;
+static unsigned short kcal_hue = 0;
+static unsigned short kcal_sat = 255;
+static unsigned short kcal_val = 255;
+static unsigned short kcal_cont = 255;
+
+module_param(kcal_red, short, 0644);
+module_param(kcal_green, short, 0644);
+module_param(kcal_blue, short, 0644);
+module_param(kcal_hue, short, 0644);
+module_param(kcal_sat, short, 0644);
+module_param(kcal_val, short, 0644);
+module_param(kcal_cont, short, 0644);
+
 static int sde_write_3d_gamut(struct sde_hw_blk_reg_map *hw,
 		struct drm_msm_3d_gamut *payload, u32 base,
 		u32 *opcode, u32 pipe, u32 scale_tbl_a_len,
@@ -210,7 +228,9 @@ void sde_setup_dspp_pccv4(struct sde_hw_dspp *ctx, void *cfg)
 	struct drm_msm_pcc *pcc_cfg;
 	struct drm_msm_pcc_coeff *coeffs = NULL;
 	int i = 0;
+	int kcal_min = 20;
 	u32 base = 0;
+        u32 opcode = 0, local_opcode = 0;
 #ifdef OPLUS_BUG_STABILITY
 	static struct drm_msm_pcc *pcc_cfg_last;
 	struct dsi_display *display = get_main_display();
@@ -222,6 +242,13 @@ void sde_setup_dspp_pccv4(struct sde_hw_dspp *ctx, void *cfg)
 		return;
 	}
 
+	if (kcal_red < kcal_min)
+		kcal_red = kcal_min;
+	if (kcal_green < kcal_min)
+		kcal_green = kcal_min;
+	if (kcal_blue < kcal_min)
+		kcal_blue = kcal_min;
+
 	if (!hw_cfg->payload) {
 		DRM_DEBUG_DRIVER("disable pcc feature\n");
 		SDE_REG_WRITE(&ctx->hw, ctx->cap->sblk->pcc.base, 0);
@@ -279,15 +306,48 @@ void sde_setup_dspp_pccv4(struct sde_hw_dspp *ctx, void *cfg)
 		}
 
 		SDE_REG_WRITE(&ctx->hw, base + PCC_C_OFF, coeffs->c);
-		SDE_REG_WRITE(&ctx->hw, base + PCC_R_OFF, coeffs->r);
-		SDE_REG_WRITE(&ctx->hw, base + PCC_G_OFF, coeffs->g);
-		SDE_REG_WRITE(&ctx->hw, base + PCC_B_OFF, coeffs->b);
+
+		// RED
+		SDE_REG_WRITE(&ctx->hw, base + PCC_R_OFF,
+			i == 0 ? (coeffs->r * kcal_red) / 256 : coeffs->r);
+		// GREEN
+		SDE_REG_WRITE(&ctx->hw, base + PCC_G_OFF,
+			i == 1 ? (coeffs->g * kcal_green) / 256 : coeffs->g);
+		// BLUE
+		SDE_REG_WRITE(&ctx->hw, base + PCC_B_OFF,
+			i == 2 ? (coeffs->b * kcal_blue) / 256 : coeffs->b);
+
 		SDE_REG_WRITE(&ctx->hw, base + PCC_RG_OFF, coeffs->rg);
 		SDE_REG_WRITE(&ctx->hw, base + PCC_RB_OFF, coeffs->rb);
 		SDE_REG_WRITE(&ctx->hw, base + PCC_GB_OFF, coeffs->gb);
 		SDE_REG_WRITE(&ctx->hw, base + PCC_RGB_OFF, coeffs->rgb);
 	}
 
+	opcode = SDE_REG_READ(&ctx->hw, ctx->cap->sblk->hsic.base);
+
+	// HUE
+	SDE_REG_WRITE(&ctx->hw, ctx->cap->sblk->hsic.base + PA_HUE_OFF,
+		kcal_hue & PA_HUE_MASK);
+	local_opcode |= PA_HUE_EN;
+
+	// SATURATION
+	SDE_REG_WRITE(&ctx->hw, ctx->cap->sblk->hsic.base + PA_SAT_OFF,
+		kcal_sat & PA_SAT_MASK);
+	local_opcode |= PA_SAT_EN;
+
+	// VALUE
+	SDE_REG_WRITE(&ctx->hw, ctx->cap->sblk->hsic.base + PA_VAL_OFF,
+		kcal_val & PA_VAL_MASK);
+	local_opcode |= PA_VAL_EN;
+
+	// CONTRAST
+	SDE_REG_WRITE(&ctx->hw, ctx->cap->sblk->hsic.base + PA_CONT_OFF,
+		kcal_cont & PA_CONT_MASK);
+	local_opcode |= PA_CONT_EN;
+
+	opcode |= (local_opcode | PA_EN);
+	SDE_REG_WRITE(&ctx->hw, ctx->cap->sblk->hsic.base, opcode);
+
 	SDE_REG_WRITE(&ctx->hw, ctx->cap->sblk->pcc.base, PCC_EN);
 #ifdef OPLUS_BUG_STABILITY
 	if (display != NULL && display->panel != NULL) {
diff --git a/techpack/display/msm/sde/sde_hw_dspp.c b/techpack/display/msm/sde/sde_hw_dspp.c
index 511fb720e801..77411da872cf 100644
--- a/techpack/display/msm/sde/sde_hw_dspp.c
+++ b/techpack/display/msm/sde/sde_hw_dspp.c
@@ -65,7 +65,7 @@ static void dspp_pcc(struct sde_hw_dspp *c)
 		c->ops.setup_pcc = sde_setup_dspp_pcc_v1_7;
 	else if (c->cap->sblk->pcc.version ==
 			(SDE_COLOR_PROCESS_VER(0x4, 0x0))) {
-		ret = reg_dmav1_init_dspp_op_v4(SDE_DSPP_PCC, c->idx);
+		ret = 1; // reg_dmav1_init_dspp_op_v4(SDE_DSPP_PCC, c->idx); use the sde one instead, with correct hsic settings possibilities
 #ifdef OPLUS_BUG_STABILITY
 		if (!ret)
 			c->ops.setup_pcc = sde_setup_dspp_pccv4;
diff --git a/techpack/display/msm/sde/sde_kms.c b/techpack/display/msm/sde/sde_kms.c
index ba293cd8f309..e4ee3e14c868 100644
--- a/techpack/display/msm/sde/sde_kms.c
+++ b/techpack/display/msm/sde/sde_kms.c
@@ -5192,3 +5192,26 @@ int sde_kms_handle_recovery(struct drm_encoder *encoder)
 	SDE_EVT32(DRMID(encoder), MSM_ENC_ACTIVE_REGION);
 	return sde_encoder_wait_for_event(encoder, MSM_ENC_ACTIVE_REGION);
 }
+
+void sde_kms_trigger_early_wakeup(struct sde_kms *sde_kms,
+		struct drm_crtc *crtc)
+{
+	struct msm_drm_private *priv;
+	struct drm_encoder *drm_enc;
+
+	if (!sde_kms || !crtc) {
+		SDE_ERROR("invalid argument sde_kms %pK crtc %pK\n",
+			sde_kms, crtc);
+		return;
+	}
+
+	priv = sde_kms->dev->dev_private;
+
+	SDE_ATRACE_BEGIN("sde_kms_trigger_early_wakeup");
+	drm_for_each_encoder_mask(drm_enc, crtc->dev, crtc->state->encoder_mask)
+		sde_encoder_trigger_early_wakeup(drm_enc);
+
+	if (sde_kms->first_kickoff)
+		sde_power_scale_reg_bus(&priv->phandle, VOTE_INDEX_HIGH, false);
+	SDE_ATRACE_END("sde_kms_trigger_early_wakeup");
+}
diff --git a/techpack/display/msm/sde/sde_kms.h b/techpack/display/msm/sde/sde_kms.h
index 117c9748ffee..1405defeaaf4 100644
--- a/techpack/display/msm/sde/sde_kms.h
+++ b/techpack/display/msm/sde/sde_kms.h
@@ -766,4 +766,13 @@ int sde_kms_vm_trusted_prepare_commit(struct sde_kms *sde_kms,
  */
 int sde_kms_vm_primary_prepare_commit(struct sde_kms *sde_kms,
 					   struct drm_atomic_state *state);
+
+/**
+ * sde_kms_trigger_early_wakeup - trigger early wake up
+ * @sde_kms: pointer to sde_kms structure
+ * @crtc: pointer to drm_crtc structure
+ */
+void sde_kms_trigger_early_wakeup(struct sde_kms *sde_kms,
+		struct drm_crtc *crtc);
+
 #endif /* __sde_kms_H__ */
diff --git a/techpack/display/oplus/oplus_aod.c b/techpack/display/oplus/oplus_aod.c
index 2c327b812572..952042415244 100644
--- a/techpack/display/oplus/oplus_aod.c
+++ b/techpack/display/oplus/oplus_aod.c
@@ -12,7 +12,7 @@
 
 #include "dsi_defs.h"
 #include "oplus_aod.h"
-int aod_light_mode = 0;
+int aod_light_mode = 1;
 DEFINE_MUTEX(oplus_aod_light_mode_lock);
 
 int __oplus_display_set_aod_light_mode(int mode)
